IPv6 support

elcallio opened the pull request:
Fixes #2027

Modifies inet address type in scylla to use seastar::net::inet_address,
and removes explicit use of ipv4_addr in various network code in favour
of socket_address. Thus capable of resolving and binding to ipv6.

Adds config option to enable/disable ipv6 (default enabled), so
upgrading cluster can continue to work while running mixed version
nodes (since gossip message address serialization becomes different).

v2:
* Made serialization of gms::inet_address ipv6 aware (encoding
  addresses so that ipv4 encoding remains the same).
* Made address lookup family optional
* Added enable/disable option
* Removed legacy "raw" access

v3:
* Rebased on master
* Renamed enable_ipv6 to enable_ipv6_dns_lookup

avikivity commented:
@elcallio I still think default enabled is too dangerous

elcallio commented:
Then I misunderstood your comment in the email thread. But ok...

avikivity commented:
Dequeuing. This causes many failures in next promotion dtest (example I checked is user_types_test). The series does not bisect; the offending commit is one of:

3116423f9d8e1a85d18365be8208f761b80831ca
58fcd795539384b3d3c423f9eb1fd7526b118222
5f251a841b960826020ef0f7feac7aca9865f861
0fc34170b04e53415658e0c8d82ffdf84daaee75
edae6a8974ab7d8440a223f2671ed6bbe74a5299
386c8e4619151a3ba18085d996649ccdc57146e3


Please fix, make it bisectable, and add a unit test that would have caught the bug earlier.

elcallio commented:
Yeah, my bad. Managed to rebase a few lines wrong. And borked the serialization change. New version coming soon.

elcallio commented:
New version up: 
* Fixed code split error. Now builds and runs in every commit.
* Fixed serialization error. Dtests work as expected now.
* Added small test for serialization.

avikivity commented:
@elcallio can you point me at the fix? to guide my re-review.

elcallio commented:
5fd811e - templating of update_peer_info needs to exist here already
c540e36 - serialization should perhaps be same endian all the way. Fixup of eariler comment only done half way previously. 
5dfc356 - test
